File: portal/priv/static/assets/wire.js
// relay wire codecs — the browser mirror of internal/protocol/wire.go.
// Message-kind numbers are part of the contract (see PROTOCOL.md) and must match Go.

export const Kind = {
  Hello: 0x01,
  Output: 0x02,
  Exit: 0x03,
  Control: 0x04,
  Input: 0x10,
  Resize: 0x11,
  CtrlReq: 0x12,
  CtrlRel: 0x13,
};

export const Control = { ReadOnly: 0, Granted: 1, Taken: 2 };

const enc = new TextEncoder();
const dec = new TextDecoder();
const VIEWER_PAYLOAD_MAGIC = enc.encode("OVP1");

export function decodeHello(b) {
  if (b.length < 12) throw new Error("short hello");
  const dv = new DataView(b.buffer, b.byteOffset, b.byteLength);
  // The protocol's sequence numbers are uint64; JS represents them as Number, which
  // is exact only up to 2^53 (Number.MAX_SAFE_INTEGER). A terminal session would
  // need ~9 quadrillion frames to reach that, so the assumption never bites in
  // practice — see PROTOCOL.md. Same conversion in crypto.js seal/open.
  return { baseline: Number(dv.getBigUint64(0)), cols: dv.getUint16(8), rows: dv.getUint16(10) };
}

export function encodeResize(cols, rows) {
  const b = new Uint8Array(4);
  const dv = new DataView(b.buffer);
  dv.setUint16(0, cols);
  dv.setUint16(2, rows);
  return b;
}

export function decodeExit(b) {
  if (b.length < 4) throw new Error("short exit");
  return new DataView(b.buffer, b.byteOffset, b.byteLength).getInt32(0);
}

export function decodeControl(b) {
  return b.length > 0 ? b[0] : 0;
}

export function encodeViewerPayload(viewerId, payload) {
  if (!viewerId) return payload;
  const id = enc.encode(viewerId).subarray(0, 255);
  const out = new Uint8Array(VIEWER_PAYLOAD_MAGIC.length + 1 + id.length + payload.length);
  out.set(VIEWER_PAYLOAD_MAGIC, 0);
  out[VIEWER_PAYLOAD_MAGIC.length] = id.length;
  out.set(id, VIEWER_PAYLOAD_MAGIC.length + 1);
  out.set(payload, VIEWER_PAYLOAD_MAGIC.length + 1 + id.length);
  return out;
}

export function decodeViewerPayload(payload) {
  if (payload.length < VIEWER_PAYLOAD_MAGIC.length) return { viewerId: "", payload };
  for (let i = 0; i < VIEWER_PAYLOAD_MAGIC.length; i++) {
    if (payload[i] !== VIEWER_PAYLOAD_MAGIC[i]) return { viewerId: "", payload };
  }
  if (payload.length < VIEWER_PAYLOAD_MAGIC.length + 1) throw new Error("short viewer payload");
  const n = payload[VIEWER_PAYLOAD_MAGIC.length];
  const start = VIEWER_PAYLOAD_MAGIC.length + 1;
  if (payload.length < start + n) throw new Error("short viewer payload");
  return { viewerId: dec.decode(payload.subarray(start, start + n)), payload: payload.subarray(start + n) };
}

// EMPTY is the shared empty payload for control-request/release frames.
export const EMPTY = new Uint8Array(0);

File: runner/internal/protocol/wire.go
package protocol

import (
	"encoding/binary"
	"errors"
)

// Message kinds. Low byte space is runner→viewer, high is viewer→runner; see
// PROTOCOL.md. The numbers are part of the contract and must not change.
const (
	KindHello   byte = 0x01 // runner→viewer: baseline seq + initial size
	KindOutput  byte = 0x02 // runner→viewer: raw PTY output
	KindExit    byte = 0x03 // runner→viewer: command exit code
	KindControl byte = 0x04 // runner→viewer: control state (read-only / granted / taken)

	KindInput   byte = 0x10 // viewer→runner: raw keystrokes
	KindResize  byte = 0x11 // viewer→runner: terminal size
	KindCtrlReq byte = 0x12 // viewer→runner: request control
	KindCtrlRel byte = 0x13 // viewer→runner: release control
)

// Control states carried by KindControl.
const (
	ControlReadOnly byte = 0
	ControlGranted  byte = 1
	ControlTaken    byte = 2
)

var errShortPayload = errors.New("protocol: payload too short")

var (
	viewerPayloadMagic = []byte("OVP1")
	relayViewerMagic   = []byte("OTV1")
)

// EncodeHello builds a HELLO payload: the input-seq baseline the viewer must start
// at, plus the current terminal size.
func EncodeHello(baseline uint64, cols, rows uint16) []byte {
	b := make([]byte, 12)
	binary.BigEndian.PutUint64(b[0:8], baseline)
	binary.BigEndian.PutUint16(b[8:10], cols)
	binary.BigEndian.PutUint16(b[10:12], rows)
	return b
}

// DecodeHello parses a HELLO payload.
func DecodeHello(b []byte) (baseline uint64, cols, rows uint16, err error) {
	if len(b) < 12 {
		return 0, 0, 0, errShortPayload
	}
	return binary.BigEndian.Uint64(b[0:8]),
		binary.BigEndian.Uint16(b[8:10]),
		binary.BigEndian.Uint16(b[10:12]), nil
}

// EncodeResize builds a RESIZE payload.
func EncodeResize(cols, rows uint16) []byte {
	b := make([]byte, 4)
	binary.BigEndian.PutUint16(b[0:2], cols)
	binary.BigEndian.PutUint16(b[2:4], rows)
	return b
}

// DecodeResize parses a RESIZE payload.
func DecodeResize(b []byte) (cols, rows uint16, err error) {
	if len(b) < 4 {
		return 0, 0, errShortPayload
	}
	return binary.BigEndian.Uint16(b[0:2]), binary.BigEndian.Uint16(b[2:4]), nil
}

// EncodeExit builds an EXIT payload.
func EncodeExit(code int32) []byte {
	b := make([]byte, 4)
	binary.BigEndian.PutUint32(b, uint32(code))
	return b
}

// DecodeExit parses an EXIT payload.
func DecodeExit(b []byte) (int32, error) {
	if len(b) < 4 {
		return 0, errShortPayload
	}
	return int32(binary.BigEndian.Uint32(b)), nil
}

// EncodeViewerPayload binds a viewer id to an encrypted viewer→runner payload. The
// relay cannot read this wrapper, but the runner can compare it with the relay's
// plaintext source label to reject re-labeled frames.
func EncodeViewerPayload(viewerID string, payload []byte) []byte {
	if viewerID == "" {
		return payload
	}
	if len(viewerID) > 255 {
		viewerID = viewerID[:255]
	}
	b := make([]byte, len(viewerPayloadMagic)+1+len(viewerID)+len(payload))
	copy(b, viewerPayloadMagic)
	b[len(viewerPayloadMagic)] = byte(len(viewerID))
	copy(b[len(viewerPayloadMagic)+1:], viewerID)
	copy(b[len(viewerPayloadMagic)+1+len(viewerID):], payload)
	return b
}

// DecodeViewerPayload unwraps EncodeViewerPayload. Legacy payloads without the magic
// are returned as viewer id "" and the original payload.
func DecodeViewerPayload(b []byte) (viewerID string, payload []byte, err error) {
	if len(b) < len(viewerPayloadMagic) || string(b[:len(viewerPayloadMagic)]) != string(viewerPayloadMagic) {
		return "", b, nil
	}
	if len(b) < len(viewerPayloadMagic)+1 {
		return "", nil, errShortPayload
	}
	n := int(b[len(viewerPayloadMagic)])
	start := len(viewerPayloadMagic) + 1
	if len(b) < start+n {
		return "", nil, errShortPayload
	}
	return string(b[start : start+n]), b[start+n:], nil
}

// EncodeRelayViewerFrame labels an opaque encrypted viewer→runner frame with the
// relay-assigned viewer id. It carries metadata only; terminal bytes remain inside
// the sealed frame.
func EncodeRelayViewerFrame(viewerID string, sealedFrame []byte) []byte {
	if viewerID == "" {
		return sealedFrame
	}
	if len(viewerID) > 255 {
		viewerID = viewerID[:255]
	}
	b := make([]byte, len(relayViewerMagic)+1+len(viewerID)+len(sealedFrame))
	copy(b, relayViewerMagic)
	b[len(relayViewerMagic)] = byte(len(viewerID))
	copy(b[len(relayViewerMagic)+1:], viewerID)
	copy(b[len(relayViewerMagic)+1+len(viewerID):], sealedFrame)
	return b
}

// DecodeRelayViewerFrame unwraps EncodeRelayViewerFrame. Raw legacy frames are
// returned with viewer id "" and the original frame.
func DecodeRelayViewerFrame(b []byte) (viewerID string, sealedFrame []byte, err error) {
	if len(b) < len(relayViewerMagic) || string(b[:len(relayViewerMagic)]) != string(relayViewerMagic) {
		return "", b, nil
	}
	if len(b) < len(relayViewerMagic)+1 {
		return "", nil, errShortPayload
	}
	n := int(b[len(relayViewerMagic)])
	start := len(relayViewerMagic) + 1
	if len(b) < start+n {
		return "", nil, errShortPayload
	}
	return string(b[start : start+n]), b[start+n:], nil
}
